crypto/init.c

```diff
@@ -97,17 +97,19 @@ static int win32atexit(void)
 
 DEFINE_RUN_ONCE_STATIC(ossl_init_register_atexit)
 {
-#ifdef OPENSSL_INIT_DEBUG
+#if ! defined(OPENSSL_NO_DLL_ATEXIT_HANDLERS) || !OPENSSL_NO_DLL_ATEXIT_HANDLERS
```

Comment:
I would prefer it if "no-atexit" was a proper Configure option that works the same way as all the other Configure options.

Reply:
> I would prefer it if "no-atexit" was a proper Configure option that works the same way as all the other Configure options.

I had that in the original unpushed draft (`OPENSSL_NO_ATEXIT_HANDLERS`). The problem with Configure is that there does not appear to be a way to remove a **defines** entry after it is specified. My intent was to add this to all NonStop configurations, and if someone wants to keep `atexit()` (at their own risk), they could turn it off, but I could not figure out how to make that work in the Configure infrastructure, like a hypothetical **undefine** option. There are a lot of NonStop variant configurations that would each require the option instead of the base entry. If that's the only way, I'll take it. @mattcaswell please advise on this.

Reply:
Also, the only `atexit()` suppression I added was in `crypto/init.c`, All other `atexit()` situations, thoughout OpenSSL would otherwise have to be instrumented, which is not actually appropriate. In `apps`, there are `atexit()` cleanups of socket connections that are required for operability considerations - if you do not close a socket cleanly, on NonStop, it sticks around in a `FIN-WAIT` state until TTL expires, which makes the OpenSSL code much more vulnerable to DOS attacks, so I would prefer to keep this only in `crypto/init.c`. Maybe "DLL" is not the proper name, but `OPENSSL_NO_LIBRARY_ATEXIT_HANDLERS` might be better.

Reply:
I would envisage "enable-atexit" would be a Configure option and "on" by default for most platforms. Possibly there could be some platform specific code in Configure that disables at exit on NonStop *unless* "enable-atexit" has been explicitly passed at the command line.

Reply:
I'm not really comfortable with such a broad impact change. It would mean putting this into the core templates. Is that going to be acceptable? I would almost prefer to add an "undefines" function to Configure.

Reply:
> It would mean putting this into the core templates. Is that going to be acceptable?

Creating the option is a one line change in Configure (add it to the disablables array). The relevant defines are then added to configuration.h automatically. It looks also that we can automatically disable it on non-stop by modifying the non-stop template to add "atexit" to the disable setting. You shouldn't need to change the core templates.

The only remaining issue is how to "override" the disable template setting if you want to "force" the enable - but I'm sure that could be done with some code in Configure somehow.

Reply:
I'm fine with the above, but have questions:

@mattcaswell When adding "atexit" to `disablables`, I get CFLAGS and CXXFLAGS containing `-atexit`, from `Configure` which is not useful. Where is the knob to change this to something like `OPENSSL_ATEXIT_HANDLERS` when not set? Weirdly, putting `-atexit` on the command line for Configure works, but `disables => ['atexit']` does not. I'm obviously missing something. Help please.

Reply:
I think I have it worked out from a C standpoint. Adding `disable-atexit` on the command line generates `OPENSSL_NO_ATEXIT`. This part should be fine. I would still like to be able to add this to the root configuration for NonStop, but the `disables` verb does not seem to be inherited - maybe another Issue?

Reply:
>  I would still like to be able to add this to the root configuration for NonStop

Doesn't adding `disables => ['atexit']` do the right thing?

Reply:
Oh!

I wonder if @levitte has an opinion on that